### PR TITLE
TELCORE-219: Fix 8-byte payload truncation on RTP header-extension forward

### PR DESCRIPTION
**Pull request by minhtuan1407-telnyx**
## Problem

On Opus passthrough (raw-forward) legs, the B2BUA drops the last 8 bytes of the Opus payload on RTP frames that carry a one-word `0xbede` header extension (e.g. the RFC 6464 audio-level extension WebRTC clients send). The decoder then garbles the short frame → audible distortion.

## Cause

In `switch_rtp_zerocopy_read_frame`, `frame->packetlen` was set from a byte count that had the RTP header-extension size subtracted in the read path, while `frame->packet` still contains the extension. The raw-forward path transmits `packetlen` bytes from `frame->packet`, so the trailing ext-size payload bytes are cut (8 for a one-word extension).

## Fix

```c
frame->packetlen = rtp_session->last_recv_bytes;
```

`last_recv_bytes` is the full received datagram length captured before extension stripping, and is already used as the forward length on the fork/recording path. It is correct for any extension size and avoids reconstructing the length from cached per-session extension state (`recv_rtp_exts_size` is `uint8_t`, is not reset per packet, and is only assigned when the extension length field is non-zero — so it can be stale or wrapped, which over-forwards past the datagram on a zero-length extension following a non-zero one).

## Tests

`tests/unit/switch_rtp_pcap.c` — `make test-run`:
- `test_rtp_header_extension_passthrough_length`: common one-word extension (the original symptom).
- `test_rtp_header_extension_passthrough_length_edge_cases`: zero-, two-, three-word extensions and a mixed/stale-state sequence (`1→0→3→0→2→0`), asserting `frame->packetlen` equals the full wire length for each. Fails with the cached-size approach, passes with `last_recv_bytes`.

QA: 16 real-call samples (both legs), `truncated_8B = 0` across all; 6 hit the passthrough path with extension frames forwarded byte-identical.

### CODE REVIEW

**Review by damirn — APPROVED** [review]
nice find!

File: src/switch_rtp.c [hunk 1]
```diff
@@ -9673,7 +9721,7 @@ SWITCH_DECLARE(switch_status_t) switch_rtp_zerocopy_read_frame(switch_rtp_t *rtp
 	} else {
 
 		frame->packet = &rtp_session->recv_msg;
-		frame->packetlen = bytes;
+		frame->packetlen = rtp_session->last_recv_bytes;
```
Comment (dev-ryanc) [inline src/switch_rtp.c]: I think this will have an issue when fork rx/tx is enabled, it will increase a possibility that last_recv_bytes may not match with the packet/recv_msg

**Review by dev-ryanc — CHANGES_REQUESTED** [review]

File: src/switch_rtp.c (hunk 1 above)
Comment (minhtuan1407-telnyx) [reply src/switch_rtp.c]: Good catch — valid concern, I will check it. `last_recv_bytes` is session-level state that can mismatch the packet in `recv_msg` when fork rx/tx is active (it gets reassigned to `rtp_header_len + last_datalen` on the fork read path, and the jitter buffer can return a different packet too). I will update the change to derive the length from the packet actually returned instead of session state:

```c
frame->packetlen = (uint32_t)((char *)frame->data - (char *)&rtp_session->recv_msg) + (bytes - rtp_header_len);
```

That is the payload offset (header + CSRC + extension) plus the payload length of the packet currently in `recv_msg`, so it stays correct regardless of fork or JB reordering, and is consistent with `frame->datalen` by construction. Will update locally and QA it.

**Review by tajamulTelnyx — APPROVED** [review]
LGTM

**Comment by minhtuan1407-telnyx** [comment]
@dev-ryanc thanks for the catch — pushed `da47081` addressing it.

**Change:** instead of `last_recv_bytes` (session state that fork rx/tx and the jitter buffer can leave pointing at a different packet than `recv_msg`), `frame->packetlen` is now derived from the returned frame itself:

```c
frame->packetlen = (uint32_t)((char *)frame->data - (char *)&rtp_session->recv_msg) + (bytes - rtp_header_len);
```

That is the payload offset (header + CSRC + extension) plus the payload length of the packet currently in `recv_msg`, so it is correct regardless of fork or JB reordering and is consistent with `frame->datalen` by construction. Also dropped the redundant single-case test (the edge-case test already covers the one-word case).

**QA (current fix build, byte-level tshark check `out == in[:-8]` on both legs):**
- 0 truncation across all post-fix samples (now ~40 calls), both directions.
- The bug-triggering combination (passthrough leg carrying `0xbede` extension frames) was exercised and came out **byte-for-byte intact** — e.g. 1336 / 326 / 346 / 156 / 150 extension frames forwarded with 0 truncated across several calls. On the buggy build every one of those loses 8 bytes.
- Pre-fix reference reconfirmed with the same check (SQuq3wKz 828, LN17Kpon 151, original y0hFPR28 183/195 frames truncated).

Unit tests: `test_rtp_header_extension_passthrough_length_edge_cases` (0/1/2/3-word + mixed sequence) passes against the new code. Could you take another look? 🙏

**Review by dev-ryanc — APPROVED** [review]

File: src/switch_rtp.c [hunk 2]
```diff
@@ -9673,7 +9721,9 @@ SWITCH_DECLARE(switch_status_t) switch_rtp_zerocopy_read_frame(switch_rtp_t *rtp
 	} else {
 
 		frame->packet = &rtp_session->recv_msg;
-		frame->packetlen = bytes;
+		/* Full datagram length of the returned packet (payload offset + payload), derived from
+		 * the frame itself so JB reordering / fork rx/tx can't desync it like last_recv_bytes. */
+		frame->packetlen = (uint32_t)((char *)frame->data - (char *)&rtp_session->recv_msg) + (bytes - rtp_header_len);
```
Comment (baloeng) [inline src/switch_rtp.c]: This can over-count RTP packets with CSRC entries (CC > 0)

For a packet:
`12-byte RTP header + CSRC + extension + payload`

`bytes` at this point has had the RTP header-extension bytes stripped, but not the CSRC bytes, so it is effectively: `12 + CSRC + payload`

Your change results to:
```
(12 + CSRC + extension) + (CSRC + payload)
= 12 + extension + payload + 2*CSRC
```

Instead, you could derive only the missing extension length from the payload offset:
```
uint32_t csrc_len = rtp_session->recv_msg.header.cc * 4;
uint32_t payload_offset = (uint32_t)((char *)frame->data - (char *)frame->packet);
uint32_t ext_len = 0;

if (payload_offset >= rtp_header_len + csrc_len) {
 ext_len = payload_offset - rtp_header_len - csrc_len;
}

frame->packetlen = bytes + ext_len;
```
You also add a unit test with CC > 0 plus a 0xbede extension to prove packetlen remains exactly the received wire length

**Review by baloeng — CHANGES_REQUESTED** [review]

File: src/switch_rtp.c (hunk 2 above)
Comment (tajamulTelnyx) [reply src/switch_rtp.c]: @minhtuan1407-telnyx FYA

Comment (minhtuan1407-telnyx) [reply src/switch_rtp.c]: I am checking this @tajamulTelnyx

**Comment by minhtuan1407-telnyx** [comment]
@baloeng thanks — addressed in `2e159ce`.

**1. Your `packetlen` form — applied:** `frame->packetlen = bytes + ext_len`, where `ext_len` is derived from the returned frame's payload offset minus header minus CSRC. `bytes` already holds header + CSRC + payload, so adding only the extension back gives the full datagram length with no CSRC double-count.

**2. Why your form alone wasn't enough, and what else changed:** your formula assumes `bytes` has the extension stripped when `cc > 0`. It did not — `switch_rtp_handle_extensions()` was advancing `recv_msg.ebody` past the CSRC list, and `read_rtp_packet()` advanced it a *second* time, which mis-located the extension so it was never reliably stripped for `cc > 0` (a pre-existing bug). Removed the redundant advance in `handle_extensions()` (it only needs to *read* the extension, which it now locates at `body + CSRC` locally); `read_rtp_packet()`'s CSRC advance stays as the single correct one. Now the extension is stripped for `cc > 0` too, so `bytes` is consistent and your formula holds for both `cc=0` and `cc > 0`.

**3. Your test — added and passing:** `test_rtp_header_extension_passthrough_length_csrc` injects `cc=2` + a one-word `0xbede` extension and asserts `packetlen == full wire length (48)`. It fails on the previous commit (over-counts to 56) and passes now. Full `switch_rtp_pcap` suite green (3/3).

**Also covers the other review concerns:** derived from the returned frame, not `last_recv_bytes`, so it stays correct under jitter-buffer reordering / fork rx/tx (ryanc/HWR). And it fixes the original 8-byte truncation on `cc=0` passthrough + extension (TELCORE-219).

**QA (this build, byte-level `out == in[:-8]` tshark check, both legs):** 0 truncation across ~50 post-fix call samples; passthrough legs carrying `0xbede` extension frames forwarded them byte-for-byte intact (e.g. 1336 / 459 / 401 / 338 / 187 / 132 ext frames across several calls, 0 truncated) — impossible on the buggy build.

Could you take another look?

**Review by baloeng — APPROVED** [review]
